src/core/layout/qgscompositionconverter.cpp

```diff
@@ -0,0 +1,1414 @@
+  layout->setName( name );
+  // TODO: check that it is always landscape
+  QgsLayoutSize pageSize( paperWidth, paperHeight );
+  for ( int j = 0; j < pages; j++ )
+  {
+    QgsLayoutItemPage *page = QgsLayoutItemPage::create( layout );
+    page->setPageSize( pageSize );
+    layout->pageCollection()->addPage( page );
+  }
+  addItemsFromCompositionXml( layout, parentElement );
+  return layout;
+}
+
+void QgsCompositionConverter::adjustPos( QgsLayout *layout, QgsLayoutItem *layoutItem, QDomNode &itemNode, QPointF *position, bool &pasteInPlace, int zOrderOffset, QPointF &pasteShiftPos, int &pageNumber )
+{
+  if ( position )
+  {
+    if ( pasteInPlace )
+    {
+      QgsLayoutPoint posOnPage = QgsLayoutPoint::decodePoint( itemNode.toElement().attribute( QStringLiteral( "positionOnPage" ) ) );
+      layoutItem->attemptMove( posOnPage, true, false, pageNumber );
+    }
+    else
+    {
+      layoutItem->attemptMoveBy( pasteShiftPos.x(), pasteShiftPos.y() );
+    }
+  }
+
+  layout->addLayoutItem( layoutItem );
+  layoutItem->setZValue( layoutItem->zValue() + zOrderOffset );
+}
+
+void QgsCompositionConverter::restoreGeneralComposeItemProperties( QgsLayoutItem *layoutItem, const QDomElement &itemElem )
+{
+  //restore general composer item properties
+  QDomNodeList composerItemList = itemElem.elementsByTagName( QStringLiteral( "ComposerItem" ) );
+  if ( !composerItemList.isEmpty() )
+  {
+    QDomElement composerItemElem = composerItemList.at( 0 ).toElement();
+
+    //rotation
+    if ( !qgsDoubleNear( composerItemElem.attribute( QStringLiteral( "rotation" ), QStringLiteral( "0" ) ).toDouble(), 0.0 ) )
+    {
+      //check for old (pre 2.1) rotation attribute
+      layoutItem->setItemRotation( composerItemElem.attribute( QStringLiteral( "rotation" ), QStringLiteral( "0" ) ).toDouble(), false );
+    }
+
+    QgsCompositionConverter::readXml( layoutItem, composerItemElem );
+
+    // Frame color
+
+    // Background color
+  }
+}
+
+QList<QgsLayoutItem *> QgsCompositionConverter::addItemsFromCompositionXml( QgsLayout *layout, const QDomElement &parentElement, QPointF *position, bool pasteInPlace )
+{
+
+  initPropertyDefinitions();
+
+  QList< QgsLayoutItem * > newItems;
+  // Not used: QList< QgsLayoutMultiFrame * > newMultiFrames;
+
+  //if we are adding items to a layout which already contains items, we need to make sure
+  //these items are placed at the top of the layout and that zValues are not duplicated
+  //so, calculate an offset which needs to be added to the zValue of created items
+  int zOrderOffset = layout->mItemsModel->zOrderListSize();
+
+  QPointF pasteShiftPos;
+  int pageNumber = -1;
+  if ( position )
+  {
+    //If we are placing items relative to a certain point, then calculate how much we need
+    //to shift the items by so that they are placed at this point
+    //First, calculate the minimum position from the xml
+    // TODO: check this!
+    QPointF minItemPos = layout->minPointFromXml( parentElement );
```

Comment:
again, I don't think this will work with composer xml - you might need to copy the composition method for this and move it to a private member here.